### PR TITLE
Add dockerfile to containerize development

### PR DESCRIPTION
**Pull request by samsonvd**
Dockerfile reflecting circleci pipeline so the IDE can be run locally for development.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -50,8 +50,21 @@ Default credentials are set up in `config/config.ini` (`test-user:test-user` by
 
 The PHP [development server][php-web-server] can be run via `php -S localhost:8000`.
 
+### Running with Docker
+
+Install Docker, build the image, then run it.
+
+```
+docker build -t srobo-ide:local .
+docker run -d --rm --network=host -t srobo-ide:local
```
Comment (PeterJCLaw) [inline README.md:59]: As I understand it, using `--rm` here means that the docker image will be deleted when it stops, which means that any logs or content created within the IDE will disappear at that point.

Combined with the `COPY` operation which pulls a current snapshot of the code state into the image, this doesn't feel especially conducive to development to me -- the logs are useful to help diagnose things and I've found that having some repos ("projects" in IDE terms) around is useful so you don't need to start from scratch every time. Certainly I wouldn't want to have to recreate the projects for every code change made!

**Review by PeterJCLaw — COMMENTED** [review]
Thanks for the contribution. I appreciate the setup docs weren't in a great place; I've just pushed some changes which will hopefully help with that.

Personally I don't think I'd be likely to use a docker setup -- a lot of the development for the IDE happens by either writing and then running the tests (which it's not clear to me how you'd do in Docker) or by inspecting the state of the git repos which the IDE manages (which them being in Docker would rather get in the way of). I've also generally found that using Docker tends to get in the way a _lot_ more than it simplifies.

The Apache HTTPD stuff is already somewhat legacy and I'd like to move away from using Apache in deployment (I stopped using it in development a while back, in favour of the PHP development server). I've clarified this in the changes I've just pushed -- your suggestion of introducing sections there was a good one.

I also definitely _don't_ want the Docker setup to either be considered for deployment (we do have Docker in production, but it was a right pain and I'll be ripping it out as soon as I can) nor become a "de-facto" standard because it happens to match how CI currently works. The CI should perhaps make more use of the makefile, however as it's trying to validate multiple Python versions in order to aid the migration to Python 3 (something the makefile doesn't really care about) it's a bit challenging to have it do that.

**Comment by MarsCapone** [comment]
I will rearrange and fix things. I wasn't intending that things would deploy as containers, nor that it's the easiest thing to use. But I wanted to add it because I personally don't like clogging up my machine with dependencies of projects I'm working on. And it's also nice to have repeatable builds.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+RUN docker-php-ext-install gd pcntl
+
+COPY ./ /repo
+
+WORKDIR /repo
+
+RUN python --version \
+    && pip install -r pylint-requirements.txt \
+    && echo "pylint.path = $(which pylint)" >> config/automagic.ini \
+    && echo "python.path = $(which python)" >> config/automagic.ini \
```
[the remark this answers is not in the document]
Comment (MarsCapone) [reply Dockerfile:23]: I'll give it a go. Quite a lot of the makefile will automatically do steps I don't want though (as it should), so no promises.

File: README.md (hunk 1 above)
Comment (MarsCapone) [reply README.md:59]: Typically I would mount the host development files in the same place on the container, and then run a shell in the container to take different actions. But I'll document this.

File: Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (PeterJCLaw) [reply Dockerfile:23]: `make config/automagic.ini` should do exactly just these lines.

File: README.md [hunk 3]
```diff
@@ -50,8 +50,21 @@ Default credentials are set up in `config/config.ini` (`test-user:test-user` by
...
+```
+
+Then visit http://localhost:8080
```
[the remark this answers is not in the document]
Comment (MarsCapone) [reply README.md:76]: I've explained a bit more. Do you want actual step-by-step instructions of all of those things.

File: run-tests [hunk 4]
```diff
@@ -1,2 +1,3 @@
 #!/bin/bash
+make dev
```
Comment (PeterJCLaw) [inline run-tests:2]: This feels like it's going to interfere with the tests and at best combines concerns which ought to be separate. I'd really rather not do this.

File: makefile [hunk 5]
```diff
@@ -59,3 +62,14 @@ metapackages/deb/srobo-ide.deb: metapackages/deb/srobo-ide/DEBIAN/control
...
+
+# Run Docker commands
+docker-run-ide:
+	docker-compose run ide
+
+docker-develop:
+	$(DOCKER_COMPOSE_COMMAND) run dev
+
+docker-run-tests:
+	$(DOCKER_COMPOSE_COMMAND) run dev run-tests
```
Comment (PeterJCLaw) [inline makefile:75]: missing newline at the end of this file (and others)

File: local_tests_config.py [hunk 6]
```diff
@@ -0,0 +1 @@
+URL = "http://localhost:8000"
```
Comment (PeterJCLaw) [inline local_tests_config.py:1]: Can we put this tests specific content somewhere other than the repo root?

File: README.md [hunk 7]
```diff
@@ -58,6 +58,77 @@ By default you can login with any non-empty username and password.
 
 [php-web-server]: https://www.php.net/manual/en/features.commandline.webserver.php
 
+### Running with Docker
```
Comment (PeterJCLaw) [inline README.md:61]:
```suggestion
### Developing with Docker
```
I really don't want to give the idea that the IDE should be being run under Docker.

**Pull request closed without merging**